### PR TITLE
Teach command palette to ignore case when sorting items

### LINKED ISSUE #8430: Ignore case when sorting command palette
# Description of the new feature/enhancement
I took me a while to understand why custom commands I add appear at the bottom.
The reason is that when sorting commands in command palette we do it not lexicographically and as a result consider a case.

IMHO the case should not matter in the palette (or at least case-insensitive sorting should be optionally available).

# Proposed technical implementation details (optional)
Upon comparing the filtered commands, try to set both command names to lower case and then compare lexicographically.

### PR DESCRIPTION
**Pull request by Don-Vito**
Closes #8430

### CODE REVIEW

File: src/cascadia/TerminalApp/FilteredCommand.cpp [hunk 1]
```diff
@@ -229,9 +229,16 @@ namespace winrt::TerminalApp::implementation
 
         if (firstWeight == secondWeight)
         {
-            std::wstring_view firstName{ first.Command().Name() };
-            std::wstring_view secondName{ second.Command().Name() };
-            return firstName.compare(secondName) < 0;
+            const auto locale = std::locale();
```
Comment (DHowett) [inline src/cascadia/TerminalApp/FilteredCommand.cpp:232]: nit: `const std::locale locale{};`

File: src/cascadia/TerminalApp/FilteredCommand.cpp [hunk 2]
```diff
@@ -229,9 +229,16 @@ namespace winrt::TerminalApp::implementation
...
+            const auto& caseConverter = std::use_facet<std::ctype<wchar_t>>(locale);
+            const auto& lexComparator = std::use_facet<std::collate<wchar_t>>(locale);
+
+            std::wstring firstName{ first.Command().Name() };
+            std::wstring secondName{ second.Command().Name() };
+
+            caseConverter.tolower(firstName.data(), firstName.data() + firstName.size());
+            caseConverter.tolower(secondName.data(), secondName.data() + secondName.size());
+            return lexComparator.compare(firstName.data(), firstName.data() + firstName.size(), secondName.data(), secondName.data() + secondName.size()) < 0;
```
Comment (DHowett) [inline src/cascadia/TerminalApp/FilteredCommand.cpp:241]: surprised there isn't a nicer way to do this -- can we use begin/end iterators instead?

**Review by DHowett — APPROVED** [review]
no need for a setting; this is the right thing to do

Worried about creating a new locale object every time. is that expensive?

**Comment by j4james** [comment]
Note that you shouldn't need to convert the strings to lowercase if you use a proper locale. I believe the default locale uses ASCII ordering, which is why you're probably seeing things ordered as `ABC..abc...`. But if you use the user's real locale, you should get something like `aAbBcC...`.  The exact ordering will depend on the language (especially for accented characters), but that's assumedly what you want.

**Comment by Don-Vito** [comment]
> no need for a setting; this is the right thing to do
> 
> Worried about creating a new locale object every time. is that expensive?

The locale is not created every time, the runtime caches it. The major concern is a critical session, but we don't expect high contention.

File: src/cascadia/TerminalApp/FilteredCommand.cpp (hunk 2 above)
Comment (Don-Vito) [reply src/cascadia/TerminalApp/FilteredCommand.cpp:241]: We can use `front`, `back`

**Comment by Don-Vito** [comment]
> 
> 
> Note that you shouldn't need to convert the strings to lowercase if you use a proper locale. I believe the default locale uses ASCII ordering, which is why you're probably seeing things ordered as `ABC..abc...`. But if you use the user's real locale, you should get something like `aAbBcC...`. The exact ordering will depend on the language (especially for accented characters), but that's assumedly what you want.

@j4james - You are right. The problem is that right now there were few contradicting requests:
* https://github.com/microsoft/terminal/issues/6953 - localize the comparison to user settings
* https://github.com/microsoft/terminal/issues/7039 - do not localize the palette and use US-English explicitly

It seems that the team prefers the second one, but we did none of them - and thus we use an ASCII comparison. :blush: 

What I tried to do is
* Make this work now - by converting to lower case
* Add future compatibility - by reading the default locale - so once we decide to localize the comparison to user preferred language or to US-English explicitly, we will already use it.

Probably I should completely abandon the lex part for now until we decide on proper localization... 🤔

**Comment by Don-Vito** [comment]
@j4james - thanks again for advice. Using correct locale is much simpler ☺️

**Review by zadjii-msft — APPROVED** [review]
Thanks for adding a test too 🙌

**Review by rheh — APPROVED** [review]
LGMT

**Review by DHowett — APPROVED** [review]
Yup, this seems totally right. Thanks!

**Pull request merged**